### ISSUE TITLE
Inconsistent handling of ambiguity codes in reference

### ISSUE THREAD

**Issue by cmnbroad** [samtools/samtools#704]
Given a .fasta with a sequence of length 31, containing a ".":

```
>iupac  31      857152f076709b2c6067edcbaaba65c7
.aAbBcCdDgGhHkKmMnNrRsStTvVwWyY
```

running "samtools faidx" produces:

`iupac	31	49	31	32`

but running "samtools dict" yields:

```
@HD	VN:1.0	SO:unsorted
@SQ	SN:iupac	LN:30	M5:c22d7686bef274eaf5abcefc0d53885f  UR:...
```

The index says the sequence has length 31, but the dictionary says the sequence has length 30. This appears to happen if the "." is moved to the middle of the sequence; dropping it makes the index and dictionary consistent with each other.

**Comment by cmnbroad** [comment]
Here is a slightly more interesting one (reference sequence length is 23, with 3 ambiguity codes):

```
>Sheila
GCTAGCRM.gactAAAAAAAAAA
```

Index (length is 20):

`Sheila	20	8	20	21`

Dictionary (length is 22):

```
@HD	VN:1.0	SO:unsorted
@SQ	SN:Sheila	LN:22	M5:50f477b79d086c2bb887a57fcd6f508e	UR:...
```

**Comment by daviesrob** [comment]
I suspect `samtools dict` is incorrect.  It actually strips out any characters which don't match `toupper` or `tolower`.  This is probably too strict.  The code in there was actually copied from [misc/md5fa.c](https://github.com/samtools/samtools/blob/develop/misc/md5fa.c).

The [sam specification](http://samtools.github.io/hts-specs/SAMv1.pdf) says the M5 should be calculated as "MD5 checksum of the sequence in the uppercase, excluding spaces but including pads (as ‘*’s).".  This is not completely unambiguous as it doesn't say what to do with characters that are not a base or `*`.

[misc/seq_cache_populate.pl](https://github.com/samtools/samtools/blob/develop/misc/seq_cache_populate.pl) simply strips white space and puts anything left through MD5.

HTSlib's CRAM `load_ref_portion()` removes any characters outside the range `!` to `~` (i.e. the 7 bit printable ones) and puts anything left through `toupper()` (see https://github.com/samtools/htslib/blob/cc556329/cram/cram_io.c#L2318).

Could you give a pointer to what GATK does when it creates a dictionary?  It would be useful to know if it does the same thing.

We should at least ensure that `samtools dict` matches what `load_ref_portion()` does.

**Comment by cmnbroad** [comment]
Agreed, it seems like the dictionary is the incorrect one. I don't think GATK(3) had an explicit tool for doing this, but [Picard](https://github.com/broadinstitute/picard/blob/cf0348512124101aa9ac292a3ab9cdf2f2ab3d2f/src/main/java/picard/sam/CreateSequenceDictionary.java#L227-L233) and [GATK4](https://github.com/broadinstitute/gatk/blob/51360c7357f47f1ce602e0a682aab3e37047440c/src/main/java/org/broadinstitute/hellbender/tools/picard/sam/CreateSequenceDictionary.java#L141-L147) both appear to just upper case.

**Comment by daviesrob** [comment]
Yes, I'll make something to fix `dict` and `md5fa`.

**Comment by daviesrob** [comment]
I've created pull request #707 to fix this.  Could you try running your `dict` implementation on [test/dat/dict.fa](https://raw.githubusercontent.com/daviesrob/samtools/36b173e/test/dat/dict.fa) to see if you match the M5 and LN tags in [test/dat/dict.out](https://raw.githubusercontent.com/daviesrob/samtools/36b173e/test/dat/dict.out), please?  My guess is that the `chr9_bit8` entry may cause problems but everything else should be OK.

I think there would be benefit in tightening up the SAM specification to say characters with bit 8 set are stripped out given that:
1. They shouldn't be there anyway
2. It neatly avoids huge cans of worms like how to deal with the [many forms of whitespace](https://en.wikipedia.org/wiki/Whitespace_(computer_science)#Unicode) available in Unicode.
3. It's what we've been doing for ages anyway.

**Comment by cmnbroad** [comment]
I get two sequences that differ (length and md5) when I create a sequence dictionary using GATK :

@SQ	SN:chr9_whitespace	LN:326	M5:5348929eee95d1ad09288b63be4e2dc4	UR:file:/Users/cnorman/projects/gatk/dict.fa
@SQ	SN:chr9_bit8	LN:392	M5:5f1708fcef5d95c6e6b6307d461d0f1b	UR:file:/Users/cnorman/projects/gatk/dict.fa

Not surprising since we neither strip out white space nor chars with the high bit set. GATK, Picard, and htsjdk all have one or more code paths that generate md5s for sequence dictionary records, so it would be nice to get them all on the same page. We just went through a long PR to get md5 calculation for CRAM slices in htsjdk to match htslib, but we focused on IUPAC codes; we didn't even consider the high-bit-set case. The BAM part of the SAM spec in (section 4.2) says sequence bases are restricted to "=ACMGRSVTWYHKDBN", (which I read as "upper case IUPAC without '.' ") ,with other characters mapped to ‘N’.

I'm wondering if silently dropping these other chars from the m5 calc is the right thing to do, given that were retaining them in the actual sequence.

**Comment by daviesrob** [comment]
You should be looking at the [SAM specification](http://samtools.github.io/hts-specs/SAMv1.pdf) section 1.3, particularly the `@SQ` records:

`LN      Reference sequence length. Range:[1,2^31-1]` and
`M5      MD5 checksum of the sequence in the uppercase, excluding spaces but including pads (as ‘*’s)`

There's nothing about IUPAC codes in there, but you should be removing white space (and probably are in the case of line endings).  This text really needs tightening up as it ignores problems like non-printable characters and international character sets.

Currently HTSlib removes all characters outside the range `[!-~]` before calculating the length and MD5, so the values do exactly match what was used for the reference.  So far we haven't noticed any problems, presumably because we stick to the traditional ACGT and the odd IUPAC ambiguity code in our references.

Possibly we should discuss this at the file formats meeting tomorrow?

**Comment by cmnbroad** [comment]
Yes, I'll try to join that meeting (though I may join a bit late).
